src/sys/wasi/p2.rs

```diff
@@ -0,0 +1,662 @@
+                                    *fd,
+                                    socket,
+                                    *subscription,
+                                    subscription.interests.unwrap(),
+                                ));
+                            }
+                        }
+
+                        libc::tcp_socket_state_tag_t::TCP_SOCKET_STATE_CONNECTED => {
+                            if writable {
+                                states.push((
+                                    ManuallyDrop::new(unsafe {
+                                        Pollable::from_handle(
+                                            socket.state.value.connected.output_pollable,
+                                        )
+                                    }),
+                                    *fd,
+                                    socket,
+                                    *subscription,
+                                    Interest::WRITABLE,
+                                ));
+                            }
+
+                            if readable {
+                                states.push((
+                                    ManuallyDrop::new(unsafe {
+                                        Pollable::from_handle(
+                                            socket.state.value.connected.input_pollable,
+                                        )
+                                    }),
+                                    *fd,
+                                    socket,
+                                    *subscription,
+                                    Interest::READABLE,
+                                ));
+                            }
+                        }
+
+                        _ => return Err(io::Error::from_raw_os_error(libc::EBADF)),
+                    }
+                }
+
+                _ => return Err(io::Error::from_raw_os_error(libc::EBADF)),
+            }
+        }
+
+        let mut pollables = states
+            .iter()
+            .map(|(pollable, ..)| pollable.deref())
+            .collect::<Vec<_>>();
+
+        let timeout = timeout.map(|timeout| {
+            monotonic_clock::subscribe_duration(timeout.as_nanos().try_into().unwrap())
+        });
+        pollables.extend(&timeout);
+
+        #[cfg(debug_assertions)]
+        if pollables.is_empty() {
+            warn!("calling mio::Poll::poll with empty pollables; this likely not what you want");
+        }
+
+        for index in poll::poll(&pollables) {
+            let index = usize::try_from(index).unwrap();
+            if timeout.is_none() || index != pollables.len() - 1 {
+                let (_, fd, socket, subscription, interests) = &states[index];
+
+                let mut push_event = || {
+                    events.push(Event {
+                        token: subscription.token,
+                        interests: *interests,
+                    })
+                };
```

Comment:
Just out of curiosity, if there any reason to do that in a closure vs. doing it inline at L276, L289 and L297?

Reply:
No particular reason other than avoiding the repetition.

Reply:
Alright! Was not sure if it was DRY or there was a mechanism I am not aware of, thanks for clarification 🙏